test/ShallowWrapper-spec.js

```diff
@@ -158,6 +158,19 @@ describe('shallow', () => {
       expect(wrapper.contains(passes2)).to.equal(true);
     });
 
+    it('shouldn\'t confuse non React elements and string literals in DOM', () => {
+      const functionalComponent = () => <div></div>;
+
+      const wrapper = shallow(
+        <div>
+          <functionalComponent />
+          { 'Text node' }
+        </div>
+      );
+
+      expect(wrapper.contains(functionalComponent)).to.equal(false);
```

Comment:
@ljharb part of the JSX spec is that lowercase component names (ie, `<functionalComponent>` should be interpreted as string "DOM" components).